### ISSUE TITLE
Accessibility issues

### ISSUE THREAD

**Issue by equinusocio** [tajo/react-range#26]
- Attribute `valuenow` doesn't match the `output value`. The first one eg return `54` while the output return `54.0`
- The current value is double announced since both `valuenow` and `output` are announced by SR with two different values (see first point). I suggest to add `aria-hidden` to the output element.
- The sliders thumbs should move/snudge by "10"  by pressing <kbd>shift + arrows</kbd>
- The whole component should have a flag/prop to match the OS a11y preferences:

  - The component style should be dark when users enable the dark interfaces for eyes related disabilities. Using CSS @media `prefers-color-scheme: dark`
  -  ~~The component should not use animations when user enable `Reduce motion`. Using CSS @media `prefers-reduced-motion: reduce`~~

**Comment by tajo** [comment]
Thanks for the feedback!

> * The sliders thumbs should move/snudge by "10"  by pressing shift + arrows

We use `page up` and `page down` to move by `10 x step`, based on [this article](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_slider_role#Keyboard_And_Focus)

> The component style should be dark when users enable the dark interfaces for eyes related disabilities. Using CSS @media prefers-color-scheme: dark

I guess this more up to the consumers since `react-range` doesn't come with colors but we could improve the examples, add a dark theme.

> The current value is double announced since both valuenow and output are announced by SR with two different values (see first point). I suggest to add aria-hidden to the output element.

Also, just something to fix in examples (we should), not the library itself.

> Attribute valuenow doesn't match the output value. The first one eg return 54 while the output return 54.0

That's just fixed formatting in the examples and it should not matter anyway since `5 === 5.0`

**Comment by equinusocio** [comment]
Good 👍

**Comment by kmcgrady** [comment]
Out of curiosity, do we have plans of implementing an [`aria-valuetext`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-valuetext_attribute)? Seems like we can just allow the component to accept `valuesText` to accept a similar array of strings to read out and ignore it if not specified.

Happy to work on a PR if this is a good idea. I'm not sure if there are other issues that need fixing here.

**Comment by tajo** [comment]
> Out of curiosity, do we have plans of implementing an [`aria-valuetext`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-valuetext_attribute)? Seems like we can just allow the component to accept `valuesText` to accept a similar array of strings to read out and ignore it if not specified.
> 
> Happy to work on a PR if this is a good idea. I'm not sure if there are other issues that need fixing here.

You still need to rely on number values and valuetext is just a screenreader alternative? If so, we could add an API that would be mapping function from numbers to text.

**Comment by kmcgrady** [comment]
That's correct. There are a few use cases for the text for the screen reader:

* Dates - dates are represented in integer since epoch work for a slider but cannot be read off as a date in the end.
* An array of discrete values - These can be something as simple as days of the week, colors of the rainbow, or just a set of numbers that are not a consistent step.

Either choices are fine. Either way, it's an additional prop that accomplishes the purpose.
* `valuesText` that just supplies the label to the screenreader
* `textFunc` that maps the value to a string (or coerced to a string).

**Comment by tajo** [comment]
Well, there is nothing preventing you to set `aria-valuetext` today based on the `value`. Maybe this doesn't need to be built in.

**Comment by kmcgrady** [comment]
If using it directly, yes. We currently use it in the `baseweb` slider component. I can open an issue with them to expose a similar prop and set it directly. Currently, we would have to do a search for all `[role="slider"]` and set it manually, which feels hacky and a little brittle.

It's adding more work, but I started here for two reasons:

1. There's a present issue on accessibility.
2. Because `aria-valuetext` is related to the slider that react-range is supposed to encapsulate. react-range sends all of the aria information [down here](https://github.com/tajo/react-range/blob/master/src/Range.tsx#L459)

If you think it really doesn't make sense and I should just reach out to baseweb, I can do that.

**Comment by tajo** [comment]
For BaseWeb you can just use [overrides](https://baseweb.design/guides/understanding-overrides/#state-props) which actually can be directly used as a mapping function from `$value` to `aria-valuetext`.

> Currently, we would have to do a search for all [role="slider"]

Not sure what you mean by this. You can make this change in one place for BaseWeb slider.

> There's a present issue on accessibility.

This is an optional prop, would not call it an a11y issue.

> Because aria-valuetext is related to the slider that react-range is supposed to encapsulate. react-range sends all of the aria information down here

Sure. However

- there is always a cost of adding new APIs
- this is an optional prop & somewhat rarer use-case
- the workaround is pretty straightforward and the new API would not add much value

**Comment by kmcgrady** [comment]
Gotcha. TIL! I'm glad there's an easier solution with overrides. I didn't think that existed, so I was basically manually editing the DOM to add it which felt awkward.
